File: backend/src/main/java/com/giproject/dto/matching/PageResponseDTO.java
package com.giproject.dto.matching;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import lombok.Builder;
import lombok.Getter;

//리액트에 페이지 관련 정보를 전달하는 애..
//페이지 번호에 따른 글목록, PageNatilon 정보, 이전, 다음 등의 정보,전체 페이지수, 현재 페이지 번호등등...
@Getter
public class PageResponseDTO<E> {

	private List<E> dtoList;
	private List<Integer> pageNumList;
	private PageRequestDTO pageRequestDTO;
	private boolean prev, next;
	private int totalCount, prevPage, nextPage, totalPage, current;
	
	@Builder(builderMethodName = "withAll")
	public PageResponseDTO(List<E> dtoList, PageRequestDTO pageRequestDTO, long totalCount) {
		
		this.dtoList = dtoList;
		System.out.println(dtoList.size());
		this.pageRequestDTO = pageRequestDTO;
		int end = (int) (Math.ceil(pageRequestDTO.getPage() / 10.0)) * 10;
		int start = end - 9;
		int last = (int) (Math.ceil((totalCount / (double) pageRequestDTO.getSize())));
		end = (end > last) ? last : end;
		this.prev = start > 1;
		this.next = totalCount > end * pageRequestDTO.getSize();

		this.pageNumList = IntStream.rangeClosed(start, end).boxed().collect(Collectors.toList());
		if (prev)
			this.prevPage = start - 1;
		if (next)
			this.nextPage = end + 1;
		this.totalPage = this.pageNumList.size();
		this.current = pageRequestDTO.getPage();
	}

}

File: flutter/flutterproject/lib/DTO/PageResponseDTO.dart
// lib/DTO/page_response_dto.dart
typedef ItemParser<T> = T Function(Map<String, dynamic>);

class PageResponseDTO<T> {
  final List<T> dtoList;
  final List<int> pageNumList;
  final bool prev;
  final bool next;
  final int totalCount;
  final int prevPage;
  final int nextPage;
  final int totalPage;
  final int current;

  PageResponseDTO({
    required this.dtoList,
    required this.pageNumList,
    required this.prev,
    required this.next,
    required this.totalCount,
    required this.prevPage,
    required this.nextPage,
    required this.totalPage,
    required this.current,
  });

  factory PageResponseDTO.fromJson(
    Map<String, dynamic> json,
    ItemParser<T> parseItem,
  ) {
    final list = (json['dtoList'] as List? ?? [])
        .map((e) => parseItem(Map<String, dynamic>.from(e as Map)))
        .toList();

    return PageResponseDTO<T>(
      dtoList: list,
      pageNumList:
          (json['pageNumList'] as List? ?? []).map((e) => int.parse('$e')).toList(),
      prev: json['prev'] == true,
      next: json['next'] == true,
      totalCount: int.tryParse('${json['totalCount']}') ?? 0,
      prevPage: int.tryParse('${json['prevPage']}') ?? 0,
      nextPage: int.tryParse('${json['nextPage']}') ?? 0,
      totalPage: int.tryParse('${json['totalPage'] ?? json['tatalPage']}') ?? 0,
      current: int.tryParse('${json['current']}') ?? 0,
    );
  }
}
